### PR TITLE
Add translate::Borrowed wrapper struct for from_glib_borrow()

### PR DESCRIPTION
**Pull request by sdroege**
Fixes https://github.com/gtk-rs/glib/issues/599

@EPashkin @GuillaumeGomez Any comments/opinions?

### CODE REVIEW

File: src/translate.rs [hunk 1]
```diff
@@ -184,6 +184,40 @@ pub struct StashMut<'a, P: Copy, T: ?Sized>(pub P, pub <T as ToGlibPtrMut<'a, P>
...
 
+#[derive(Debug)]
+pub struct Borrowed<T>(Option<T>);
+
+impl<T> Borrowed<T> {
+    pub fn new(val: T) -> Self {
+        Self(Some(val))
+    }
+
+    pub unsafe fn into_inner(mut self) -> T {
```
Comment (EPashkin) [inline src/translate.rs]: This IMHO need comment about limits of usage this function

Comment (EPashkin) [reply src/translate.rs]: Or maybe better instead add `from_borrowed` constructor?

**Comment by GuillaumeGomez** [comment]
No obvious issues that I can see...

File: src/translate.rs (hunk 1 above)
Comment (sdroege) [reply src/translate.rs]: How would `from_borrowed()` work? I was thinking of that but it didn't seem like an improvement with what I came up with :)

Docs will definitely come if the approach is considered ok

**Comment by sdroege** [comment]
> No obvious issues that I can see...

Any other objections? :)

**Comment by GuillaumeGomez** [comment]
I'll need to reread the PR completely and more carefully and I'll be rested.

File: src/translate.rs [hunk 2]
```diff
@@ -184,6 +184,40 @@ pub struct StashMut<'a, P: Copy, T: ?Sized>(pub P, pub <T as ToGlibPtrMut<'a, P>
 where
     T: ToGlibPtrMut<'a, P>;
 
+#[derive(Debug)]
+pub struct Borrowed<T>(Option<T>);
```
Comment (sdroege) [inline src/translate.rs]:
```suggestion
pub struct Borrowed<T>(mem::ManuallyDrop<T>);
```

This is better, it does not involve the annoying `Option` wrapping

Comment (EPashkin) [reply src/translate.rs]: What will be with `Borrowed::new(None)` for `Option` ?

File: src/translate.rs (hunk 1 above)
Comment (EPashkin) [reply src/translate.rs]: I think about something like `fn from_borrowed<E, F: FnOnce(E) -> T>(other: Borrowed<E>, f:F)-> Self`
With call
```
                $crate::translate::Borrowed::from_borrowed(
$crate::translate::from_glib_borrow::<_, $crate::boxed::Boxed<_, _>>(ptr),
          |v|   $name( v )
                )
```
If this don't work the :+1: for `into_inner`

Comment (EPashkin) [reply src/translate.rs]: or we can just add `fn map<E, F: FnOnce(T) -> E>(self, f:F)-> Borrowed<E>`

File: src/translate.rs (hunk 2 above)
Comment (sdroege) [reply src/translate.rs]: I don't understand :) In that case `T` is `Option<_>` and all is good

Comment (EPashkin) [reply src/translate.rs]: Really, I missed that :(

File: src/translate.rs (hunk 1 above)
Comment (sdroege) [reply src/translate.rs]: That's all doable but does not really improve the code and also does not make it any safer :) You still need to ensure that the value passed to the closure is not dropped anywhere.

I'd keep it as is.

**Comment by sdroege** [comment]
Added some docs and cleanup. Any other comments? Otherwise I would start updating everything else and gir

**Comment by EPashkin** [comment]
@sdroege Thanks again and 👍

**Comment by sdroege** [comment]
I have an initial changeset for gir (easier than expected), and gio is compiling again. Will create PRs once everything's compiling.

**Comment by sdroege** [comment]
This would be ready now but don't merge yet (want to prepare the other PRs first).

File: src/auto/versions.txt [hunk 3]
```diff
@@ -1,2 +1,2 @@
-Generated by gir (https://github.com/gtk-rs/gir @ 83cc9ad)
+Generated by gir (https://github.com/gtk-rs/gir @ 997cbcb)
 from gir-files (https://github.com/gtk-rs/gir-files @ 22748fa)
```
Comment (EPashkin) [inline src/auto/versions.txt:2]: For be same as other maybe do gir-files update too, to 480c117

Comment (sdroege) [reply src/auto/versions.txt:2]: Done

**Comment by sdroege** [comment]
All 9 PRs in place, @GuillaumeGomez please merge and then all will be green again :)

**Comment by EPashkin** [comment]
👍 for all PRs, it builds locally for me.
Only last to do is PR against sourceview when all merged.

**Pull request merged**